NumLib/NamedFunction.h

```diff
@@ -0,0 +1,123 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#ifndef NUMLIB_NAMED_FUNCTION
+#define NUMLIB_NAMED_FUNCTION
+
+#include <string>
+#include <vector>
+#include <functional>
+#include <memory>
+#include <cassert>
+#include <type_traits>
+
+namespace detail
+{
+//! Checks if all types \c Ts are the same as \c TRef.
+template<typename TRef, typename... Ts>
+struct AllTypesSameAs;
+
+template<typename TRef, typename T1, typename... Ts>
+struct AllTypesSameAs<TRef, T1, Ts...>
+{
+    static const bool value = std::is_same<TRef, T1>::value
+        && AllTypesSameAs<TRef, Ts...>::value;
+};
+
+template<typename TRef>
+struct AllTypesSameAs<TRef>
+{
+    static const bool value = true;
+};
+
+template<typename TRef, typename T1, typename... Ts>
+const bool AllTypesSameAs<TRef, T1, Ts...>::value;
+
+template<typename TRef>
+const bool AllTypesSameAs<TRef>::value;
+
+} // namespace detail
+
+
+
+namespace NumLib
+{
+//! Maximum number of function arguments supported by NamedFunction.
+const int MAX_FUNCTION_ARGS = 32;
```

Comment:
why not define this within NamedFunction class? The variable name is too general to be placed under NumLib namespace